pkgs/development/compilers/llvm/15/libcxx/default.nix

```diff
@@ -71,7 +71,8 @@ stdenv.mkDerivation rec {
       "-DLIBCXX_ENABLE_THREADS=OFF"
       "-DLIBCXX_ENABLE_FILESYSTEM=OFF"
       "-DLIBCXX_ENABLE_EXCEPTIONS=OFF"
-    ] ++ lib.optional (!enableShared) "-DLIBCXX_ENABLE_SHARED=OFF";
+    ] ++ lib.optional (!enableShared) "-DLIBCXX_ENABLE_SHARED=OFF"
+    ++ lib.optional headersOnly "-DLIBCXX_ENABLE_ABI_LINKER_SCRIPT=OFF";
```

Comment:
I'd move this line closer to the optional for `!headersOnly`, just for the sake of the reader.

Reply:
Also, I think my change here is a cleaner way of doing the headersOnly build?

https://github.com/NixOS/nixpkgs/pull/194781/commits/b145146c24b838c3a4c84e094a52e8236e2aab0f

Reply:
@lovesegfault thanks for the feedback!

> Also, I think my change here is a cleaner way of doing the headersOnly build?

Agreed; setting `LIBCXX_CXX_ABI` to `none` seems less brittle than strategically setting `libc++` build options.

I was concerned about what else `LIBCXX_CXX_ABI` influences but looking through the libcxx build defs again this seems fine.

Reply:
Right now I'm trying to fix `libcxxabi` which appears to be broken on non-`x86_64-linux` in both our PRs (`libc++` seems to have switching to using attribute `using_if_exists` to gate imports in libc++ on platforms with "incomplete" libc impls; this requires us to use a newer compiler to build `libcxx` and `libcxxabi`); I'll revise the PR with your suggestions afterwards.

Reply:
Is the build on Darwin using the system clang version? It should probably use `llvmPackages_{n-1}`, I think.

Reply:
It's currently using clang 11 (what [`llvmPackages` on darwin currently defaults to](https://github.com/NixOS/nixpkgs/blob/0bfabe2e11e091a4d37fca285e5f2a3cc45e01f2/pkgs/top-level/all-packages.nix#L14319) which the stdenv picks up).

Using `llvmPackages_14`'s stdenv (either for the full package set or just `libcxx{,abi}`) would definitely fix this issue but I'm a little wary to do so just because it's a departure from what's been done in the past and I'm not sure about the precedent it sets. In general are we concerned about having long dependency chains for packages like these?

Also fwiw, `aarch64-linux` also seems to be affected; same issue.

---

I've been trying to use the `clang` built within the package set to build `libcxxabi` instead of using a "special" stdenv. This is a build configuration `libc++` seems to support (the docs call it a ["bootstrapping build"](https://libcxx.llvm.org/BuildingLibcxx.html#bootstrapping-build)) albeit seemingly not when clang is built "separately". I'm having a little trouble convincing CMake to skip some of it's checks for a working C++ compiler; just need to spend some more time puzzling through the LLVM runtimes/libcxx build defs and find the right knobs I think.

If the dependency chain thing is a non-issue though, I'm happy to go that route instead.

Reply:
Oh, definitely just use `llvmPackages_14` instead of trying to shove new-clang in another stdenv.

Other folks can chime in: @Ericson2314 @dtzWill @matthewbauer

Reply:
@lovesegfault it actually ended up being pretty straight-forward; I just needed a CMake flag: https://github.com/NixOS/nixpkgs/pull/194634/commits/ab9d3df63862da93d8a229ea379efb7b50e8692b

---

I'm gonna mark this as resolved for now but if using `llvmPackages_14.libcxxStdenv` is preferable I'm happy to make that change